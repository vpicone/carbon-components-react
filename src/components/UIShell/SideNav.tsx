/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { settings } from 'carbon-components';
import cx from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';
import { AriaLabelPropType } from '../../prop-types/AriaPropTypes';
import SideNavFooter from './SideNavFooter';

const { prefix } = settings;

const translations = {
  'carbon.sidenav.state.open': 'Close',
  'carbon.sidenav.state.closed': 'Open',
};

function translateById(id) {
  return translations[id];
}

export default class SideNav extends React.Component {
  static propTypes = {
    /**
     * Required props for accessibility label on the underlying menu
     */
    ...AriaLabelPropType,

    /**
     * Optionally provide a custom class to apply to the underlying <li> node
     */
    className: PropTypes.string,

    /**
     * Provide a custom function for translating all message ids within this
     * component. This function will take in two arguments: the mesasge Id and the
     * state of the component. From this, you should return a string representing
     * the label you want displayed or read by screen readers.
     */
    translateById: PropTypes.func,
  };

  static defaultProps = {
    translateById,
  };

  state = {
    isExpanded: true,
  };

  handleExpand = () => {
    this.setState(state => ({ isExpanded: !state.isExpanded }));
  };

  render() {
    const {
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      children,
      className: customClassName,
      translateById: t,
    } = this.props;
    const { isExpanded } = this.state;
    const accessibilityLabel = {
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
    };
    const assistiveText = isExpanded
      ? t('carbon.sidenav.state.open')
      : t('carbon.sidenav.state.closed');
    const className = cx({
      [`${prefix}--side-nav`]: true,
      [`${prefix}--side-nav--expanded`]: isExpanded,
      [customClassName]: !!customClassName,
    });

    return (
      <nav
        className={`${prefix}--side-nav__navigation ${className}`}
        {...accessibilityLabel}>
        {children}
        <SideNavFooter
          assistiveText={assistiveText}
          isExpanded={isExpanded}
          onToggle={this.handleExpand}
        />
      </nav>
    );
  }
}
