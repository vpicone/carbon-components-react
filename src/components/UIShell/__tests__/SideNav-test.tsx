/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import SideNav from '../SideNav';

describe('SideNav', () => {
  let mockProps;

  beforeEach(() => {
    mockProps = {
      'aria-label': 'Navigation',
      children: <h2>Navigation</h2>,
    };
  });

  it('should render', () => {
    const wrapper = mount(<SideNav {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should toggle the menu expansion state when clicking on the footer', () => {
    const wrapper = mount(<SideNav {...mockProps} />);
    expect(wrapper.state('isExpanded')).toBe(true);
    wrapper.find('button').simulate('click');
    expect(wrapper.state('isExpanded')).toBe(false);
    expect(wrapper).toMatchSnapshot();
  });
});
