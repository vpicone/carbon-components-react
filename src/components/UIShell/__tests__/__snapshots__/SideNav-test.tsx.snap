// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNav should render 1`] = `
<SideNav
  aria-label="Navigation"
  translateById={[Function]}
>
  <aside
    className="bx--side-nav bx--side-nav--expanded"
  >
    <nav
      aria-label="Navigation"
      className="bx--side-nav__navigation"
      role="navigation"
    >
      <h2>
        Navigation
      </h2>
      <SideNavFooter
        assistiveText="Close"
        isExpanded={true}
        onToggle={[Function]}
      >
        <footer
          className="bx--side-nav__footer"
        >
          <button
            className="bx--side-nav__toggle"
            onClick={[Function]}
            title="Close"
            type="button"
          >
            <div
              className="bx--side-nav__icon"
            >
              <Close20
                height={20}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width={20}
                xmlns="http://www.w3.org/2000/svg"
              >
                <svg
                  aria-hidden={true}
                  focusable="false"
                  height={20}
                  preserveAspectRatio="xMidYMid meet"
                  style={
                    Object {
                      "willChange": "transform",
                    }
                  }
                  viewBox="0 0 32 32"
                  width={20}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4l6.6 6.6L8 22.6 9.4 24l6.6-6.6 6.6 6.6 1.4-1.4-6.6-6.6L24 9.4z"
                  />
                </svg>
              </Close20>
            </div>
            <span
              className="bx--assistive-text"
            >
              Close
            </span>
          </button>
        </footer>
      </SideNavFooter>
    </nav>
  </aside>
</SideNav>
`;

exports[`SideNav should toggle the menu expansion state when clicking on the footer 1`] = `
<SideNav
  aria-label="Navigation"
  translateById={[Function]}
>
  <aside
    className="bx--side-nav"
  >
    <nav
      aria-label="Navigation"
      className="bx--side-nav__navigation"
      role="navigation"
    >
      <h2>
        Navigation
      </h2>
      <SideNavFooter
        assistiveText="Open"
        isExpanded={false}
        onToggle={[Function]}
      >
        <footer
          className="bx--side-nav__footer"
        >
          <button
            className="bx--side-nav__toggle"
            onClick={[Function]}
            title="Open"
            type="button"
          >
            <div
              className="bx--side-nav__icon"
            >
              <ChevronRight20
                height={20}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width={20}
                xmlns="http://www.w3.org/2000/svg"
              >
                <svg
                  aria-hidden={true}
                  focusable="false"
                  height={20}
                  preserveAspectRatio="xMidYMid meet"
                  style={
                    Object {
                      "willChange": "transform",
                    }
                  }
                  viewBox="0 0 32 32"
                  width={20}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M22 16L12 26l-1.4-1.4 8.6-8.6-8.6-8.6L12 6z"
                  />
                </svg>
              </ChevronRight20>
            </div>
            <span
              className="bx--assistive-text"
            >
              Open
            </span>
          </button>
        </footer>
      </SideNavFooter>
    </nav>
  </aside>
</SideNav>
`;
